src/filesystem/SDL_filesystem.c

```diff
@@ -100,3 +100,105 @@ int SDL_GetPathInfo(const char *path, SDL_PathInfo *info)
     }
     return SDL_SYS_GetPathInfo(path, info);
 }
+
+
+static SDL_bool WildcardMatch(const char *pattern, const char *str)
+{
+    SDL_assert(str != NULL);
+    SDL_assert(pattern != NULL);
+
+    while (SDL_TRUE) {
+        const char pch = *(pattern++);
+        if (pch == '?') {
+            if (*(str++) == '\0') {
+                return SDL_FALSE;
+            }
+        } else if (pch == '*') {
+            const char nextch = *(pattern++);
+            while (SDL_TRUE) {
+                const char nextsch = *(str++);
+                if (nextsch == nextch) {
+                    break;
+                } else if (nextsch == '\0') {
+                    return SDL_FALSE;  // hit end of string without matching next pattern character.
+                }
+            }
+        } else if (pch != *(str++)) {
+            return SDL_FALSE;  // didn't match the next character.
+        } else if (pch == '\0') {
+            break;   // matched to the end of the string!
+        }
+    }
+
+    return SDL_TRUE;   // matched to the end of the string!
+}
+
+typedef struct GlobCallbackData
+{
+    const char *pattern;
+    SDL_IOStream *string_stream;
+    int num_entries;
+    size_t strings_len;
+} GlobCallbackData;
+
+static int SDLCALL GlobCallback(void *userdata, const char *dirname, const char *fname)
+{
+    GlobCallbackData *data = (GlobCallbackData *) userdata;
+
+    if (WildcardMatch(data->pattern, fname)) {
+        const size_t slen = SDL_strlen(fname) + 1;
+        if (SDL_WriteIO(data->string_stream, fname, slen) != slen) {
+            return -1;  // stop enumerating, return failure to the app.
+        }
+        data->num_entries++;
+        data->strings_len += slen;
+    }
+
+    return 1;  // keep enumerating.
+}
+
+
+char **SDL_GlobDirectory(const char *path, const char *pattern, int *count)
+{
+    if (count) {
+        *count = 0;
+    }
+
+    if (!path) {
+        SDL_InvalidParamError("path");
+        return NULL;
+    }
+
+    char **retval = NULL;
+
+    GlobCallbackData data;
+    SDL_zero(data);
+    data.pattern = pattern;
+    data.string_stream = SDL_IOFromDynamicMem();
+    if (data.string_stream) {
+        const int rc = SDL_EnumerateDirectory(path, GlobCallback, &data);
+        if (rc == 1) {
+            const size_t buflen = data.strings_len + ((data.num_entries + 1) * sizeof (char *));  // +1 for NULL terminator at end of array.
+            retval = (char **) SDL_malloc(buflen);
+            if (retval) {
+                void *strings = SDL_GetProperty(SDL_GetIOProperties(data.string_stream), SDL_PROP_IOSTREAM_DYNAMIC_MEMORY_POINTER, NULL);
+                SDL_assert(strings != NULL);
```

Comment:
Strings will be NULL if there are no matches.

Reply:
Actually, if you’re not going to keep the strings pointer yourself, it’s probably cleaner to just seek back to 0 and do a normal IO read here.